File: packages/flutter_shared/lib/tools/countdown_tool.dart
/// 倒计时工具
class CountdownTool {
  CountdownTool._();

  /// 倒计时结果
  static ({
    int days, int hours, int minutes, int seconds,
    bool isPast, String description,
  }) calculate(DateTime target) {
    final now = DateTime.now();
    final diff = target.difference(now);
    final isPast = diff.isNegative;
    final abs = diff.abs();

    final days = abs.inDays;
    final hours = abs.inHours % 24;
    final minutes = abs.inMinutes % 60;
    final seconds = abs.inSeconds % 60;

    String desc;
    if (isPast) {
      if (days > 365) {
        final y = days ~/ 365;
        final m = (days % 365) ~/ 30;
        desc = '已过去 $y 年 $m 个月';
      } else if (days > 30) {
        final m = days ~/ 30;
        final d = days % 30;
        desc = '已过去 $m 个月 $d 天';
      } else if (days > 0) {
        desc = '已过去 $days 天 $hours 小时';
      } else if (hours > 0) {
        desc = '已过去 $hours 小时 $minutes 分钟';
      } else if (minutes > 0) {
        desc = '已过去 $minutes 分钟 $seconds 秒';
      } else {
        desc = '刚刚过去';
      }
    } else {
      if (days > 365) {
        final y = days ~/ 365;
        final m = (days % 365) ~/ 30;
        desc = '还剩 $y 年 $m 个月';
      } else if (days > 30) {
        final m = days ~/ 30;
        final d = days % 30;
        desc = '还剩 $m 个月 $d 天';
      } else if (days > 0) {
        desc = '还剩 $days 天 $hours 小时';
      } else if (hours > 0) {
        desc = '还剩 $hours 小时 $minutes 分钟';
      } else if (minutes > 0) {
        desc = '还剩 $minutes 分钟 $seconds 秒';
      } else {
        desc = '即将到来';
      }
    }

    return (
      days: days,
      hours: hours,
      minutes: minutes,
      seconds: seconds,
      isPast: isPast,
      description: desc,
    );
  }
}

File: miniapp/src/utils/tools/countdownTool.ts
/**
 * 倒计时工具
 * 移植自 flutter_shared/tools/countdown_tool.dart
 */

export interface CountdownResult {
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
  isPast: boolean;
  description: string;
}

export const CountdownTool = {
  calculate(target: Date): CountdownResult {
    const now = new Date();
    const diff = target.getTime() - now.getTime();
    const isPast = diff < 0;
    const abs = Math.abs(diff);

    const totalMs = abs;
    const days = Math.floor(totalMs / (1000 * 60 * 60 * 24));
    const hours = Math.floor((totalMs % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
    const minutes = Math.floor((totalMs % (1000 * 60 * 60)) / (1000 * 60));
    const seconds = Math.floor((totalMs % (1000 * 60)) / 1000);

    let description: string;
    const prefix = isPast ? '已过去' : '还剩';

    if (days > 365) {
      const y = Math.floor(days / 365);
      const m = Math.floor((days % 365) / 30);
      description = `${prefix} ${y} 年 ${m} 个月`;
    } else if (days > 30) {
      const m = Math.floor(days / 30);
      const d = days % 30;
      description = `${prefix} ${m} 个月 ${d} 天`;
    } else if (days > 0) {
      description = `${prefix} ${days} 天 ${hours} 小时`;
    } else if (hours > 0) {
      description = `${prefix} ${hours} 小时 ${minutes} 分钟`;
    } else if (minutes > 0) {
      description = `${prefix} ${minutes} 分钟 ${seconds} 秒`;
    } else {
      description = isPast ? '刚刚过去' : '即将到来';
    }

    return { days, hours, minutes, seconds, isPast, description };
  },
};
